### PR TITLE
Prevent  deadlocking event-handler on missing recordings

### PR DESCRIPTION
**Pull request by rosstimothy**
If the number of session.upload events without a corresponding session recording is higher than the configured concurrency, then the event-handler could deadlock while attempting to find said recording. To prevent this from occurring processing of session recordings has been modified such that:

- if a recording does not exist and the session.upload event is from more than 48 hours in the past it is assumed to be lost and no more attempts to process the recording will happen
- all processing of recordings that were not found is now done in a separate background routine, instead of inline with event processing

The storage directory will now contain information about sessions with a missing recording. The background routine will process them periodically, capping the number of attempts per recording at 3.

Changelog: reduce the probability that the event-handler deadlocks when encountering errors processing session recordings

### CODE REVIEW

File: integrations/event-handler/session_events_job.go [hunk 1]
```diff
@@ -147,6 +119,119 @@ func (j *SessionEventsJob) run(ctx context.Context) error {
...
+		// session recording is assumed to be lost forever.
+		eventTimeCutOff = -48 * time.Hour
+	)
+
+	log := logger.Get(ctx).WithField("id", s.ID).WithField("index", s.Index)
+	backoff := backoff.NewDecorr(sessionBackoffBase, sessionBackoffMax, clockwork.NewRealClock())
+	attempt := sessionBackoffNumTries
+	retry, err := j.consumeSession(ctx, s)
+
+	for {
```
Comment (tigrato) [inline integrations/event-handler/session_events_job.go:138]:
```suggestion

	for {
		retry, err := j.consumeSession(ctx, s)
```

should this be moved into the for loop? 
otherwise retry and err will be static and we will keep ranging the same results until the break condition is triggered

File: integrations/event-handler/session_events_job.go [hunk 2]
```diff
@@ -147,6 +119,119 @@ func (j *SessionEventsJob) run(ctx context.Context) error {
...
+		// session recording will be processed before assuming the recording
+		// is lost forever.
+		maxNumberOfProcessingAttempts = 2
+		// eventTimeCutOff is the point of time in the past at which a missing
+		// session recording is assumed to be lost forever.
+		eventTimeCutOff = -48 * time.Hour
+	)
+
+	log := logger.Get(ctx).WithField("id", s.ID).WithField("index", s.Index)
+	backoff := backoff.NewDecorr(sessionBackoffBase, sessionBackoffMax, clockwork.NewRealClock())
```
Comment (tigrato) [inline integrations/event-handler/session_events_job.go:134]: should we reduce the `sessionBackoffMax` and decrease the `sessionBackoffNumTries`?
Otherwise this can lock 10min while we can continue with other sessions in the meantime

File: integrations/lib/signals.go [hunk 3]
```diff
@@ -41,6 +41,7 @@ func ServeSignals(app Terminable, shutdownTimeout time.Duration) {
 	signal.Notify(sigC,
 		syscall.SIGTERM, // graceful shutdown
 		syscall.SIGINT,  // graceful-then-fast shutdown
+		syscall.SIGUSR1,
```
Comment (marcoandredinis) [inline integrations/lib/signals.go:44]: Can we please add some kind of comment explaining what it does? 
```suggestion
		syscall.SIGUSR1, // log process diagnostic info
```

File: integrations/event-handler/app.go [hunk 4]
```diff
@@ -247,3 +253,89 @@ func (a *App) RegisterSession(ctx context.Context, e *TeleportEvent) {
 		log.Error("Registering session: ", err)
 	}
 }
+
+// Profile captures various Go pprof profiles and writes
+// them under the configured storage directory.
+func (a *App) Profile() {
```
Comment (marcoandredinis) [inline integrations/event-handler/app.go:259]: Two nits:
I think it should be a separate PR.
And we should probably move this to `integrations/lib/profile/` or `lib/profile/`

File: integrations/event-handler/state.go [hunk 5]
```diff
@@ -238,7 +245,91 @@ func (s *State) SetSessionIndex(id string, index int64) error {
...
+	b, err := json.Marshal(missingRecording{
+		Index:     sess.Index,
+		Attempt:   attempt,
+		Timestamp: sess.UploadTime,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := s.dv.Write(missingRecordingPrefix+sess.ID, b); err != nil {
```
Comment (marcoandredinis) [inline integrations/event-handler/state.go:259]: Should we use `WriteStream` instead and set the sync to true?
I will probably take longer but it brings the durability guarantee.

File: integrations/event-handler/state.go [hunk 6]
```diff
@@ -238,7 +245,91 @@ func (s *State) SetSessionIndex(id string, index int64) error {
...
+			return trace.Wrap(err)
+		}
+
+		var m missingRecording
+		if err := json.Unmarshal(b, &m); err != nil {
+			return trace.Wrap(err)
+		}
+
+		s := session{
+			ID:         key[len(missingRecordingPrefix):],
```
Comment (marcoandredinis) [inline integrations/event-handler/state.go:290]: I think we should extract the encode/decode operations over the `prefix+id` into helper functions

File: integrations/event-handler/state.go [hunk 7]
```diff
@@ -238,7 +245,91 @@ func (s *State) SetSessionIndex(id string, index int64) error {
...
+// IterateMissingRecordings finds any sessions with a missing recording and
+// provides them to the callback for processing.
+func (s *State) IterateMissingRecordings(callback func(s session, attempts int) error) error {
+	closeC := make(chan struct{})
+	defer close(closeC)
+	for key := range s.dv.KeysPrefix(missingRecordingPrefix, closeC) {
+		b, err := s.dv.Read(key)
+		if err != nil {
+			return trace.Wrap(err)
+		}
```
Comment (marcoandredinis) [inline integrations/event-handler/state.go:282]: Let's say this takes too long to process and someone removes a file on disk.
If that's the case, I assume we get some kind of `file not found` error. Should we continue the loop in this case (with a warning log)?

File: integrations/event-handler/state.go (hunk 5 above)
Comment (marcoandredinis) [reply integrations/event-handler/state.go:259]: Not really. Maybe we should replace everything with `WriteStream` in a follow up PR and keep this one as is for the sake of consistency.

**Review by marcoandredinis — APPROVED** [review]

**Review by tigrato — APPROVED** [review]
